File: Example/tree.go
package main

import (
	"fmt"
)

type tree struct {
	data int
	left *tree
	right *tree
}

func main()  {
	arr := []int{1,3,5,6,9,10}

	var node *tree
	fmt.Printf("node is %v\n", node)

	for i := 0; i < len(arr); i++ {
		node = insert(node, arr[i])
		fmt.Printf("node is %v\n", node)
	}
	preOrder(node)
}

func insert(node *tree, nodeVal int, ) *tree {

	if node == nil {
		return &tree{data: nodeVal}
	}

	if node.data == nodeVal {
		return node
	}
	if node.data > nodeVal {
		node.left = insert(node.left, nodeVal)
		return node
	}

	node.right = insert(node.right, nodeVal)
	return  node
}

// 前序遍历
func preOrder(node *tree)  {
	if node == nil {
		return
	}

	fmt.Println(node.data, " ")
	preOrder(node.left)
	preOrder(node.right)
}

// 中序遍历
func inOrder(node *tree)  {
	if node == nil {
		return
	}

	preOrder(node.left)
	fmt.Println(node.data, " ")
	preOrder(node.right)
}

// 后序遍历
func postOrder(node *tree)  {
	if node == nil {
		return
	}

	preOrder(node.left)
	preOrder(node.right)
	fmt.Println(node.data, " ")
}

File: Example/tree.php
<?php

// 使用非递归遍历二叉树

class tree
{
    public $data;

    public $right;

    public $left;

    public function __construct($data = 0, $right = null, $left = null)
    {
        $this->data  = $data;
        $this->right = $right;
        $this->left = $left;
    }
}

function create(): tree
{
    $arr = [7,4,13,1,15,6,3];

    $tree = new tree(array_shift($arr));
    foreach ($arr as $item) {
        $tree = insert($tree, $item);
    }

    return $tree;
}

/**
 * @param  $tree
 * @param int  $data
 * @return tree
 */
function insert($tree, int $data): tree
{
    if (empty($tree)) {
        $tree = new tree($data);
    }

    if ($tree->data == $data) {
        return $tree;
    }

    if ($tree->data > $data) {
        $tree->left = insert($tree->left, $data);
        return $tree;
    }
    $tree->right = insert($tree->right, $data);
    return  $tree;
}

$tree = create();

//辅助栈
$stack = [];
$stack[] = $tree;

while (!empty($stack)) {

    $tree = array_pop($stack);
    print_r($tree->data. "\n");

    if (!empty($tree->right)) {
        $stack[] = $tree->right;
    }

    if (!empty($tree->left)) {
        $stack[] = $tree->left;
    }
}